### PR TITLE
Snippet search

### LINKED ISSUE #489: Add a search field to filter snippets by title
This search field would sit at the top of a list of a user's snippets and the lists of org snippets

### PR DESCRIPTION
**Pull request by dane-johnson**
## Description

You can use fuzzy searching to find a snippet across all the snippets that you have access to.

## Motivation and Context

Closes #489 

## Checklist:

- [x] My code follows the code style of this project.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

## Browsers

- [ ] Chrome
- [x] Firefox
- [ ] Safari
- [ ] Mobile:
- [ ] Other:

## Screenshots (if appropriate):
screen shot 2017-05-24 at 4 16 29 pm

### CODE REVIEW

File: src/containers/CodesplainAppBar.jsx [hunk 1]
```diff
@@ -73,6 +75,19 @@ const styles = {
...
+  allSnippets[username] = userSnippets;
+  const orderedSnippets = _.reduce(allSnippets, (allList, snippets, role) => {
+    const next = _.reduce(snippets, (snipList, { language, lastEdited }, snippetTitle) => {
+      const obj = { role, snippetTitle, language, lastEdited };
+      return _.concat(snipList, obj);
+    }, []);
+    return _.concat(allList, next);
+  }, []);
+  return _.sortBy(orderedSnippets, ({ snippetTitle }) => snippetTitle);
+};
```
Comment (solkaz) [inline src/containers/CodesplainAppBar.jsx]: Can you clarify this code block?

**Comment by thebho** [comment]
I like the functionality, but something feels unfinished with the modal.  Maybe it could use a black header and an `x` like the Centred popup (Lightbox) on this sight https://wet-boew.github.io/v4.0-ci/demos/overlay/overlay-en.html

File: src/containers/CodesplainAppBar.jsx [hunk 2]
```diff
@@ -7,6 +7,7 @@ import {
 } from 'material-ui';
 import { withRouter } from 'react-router';
 import cookie from 'react-cookie';
+import _ from 'lodash';
```
Comment (dane-johnson) [inline src/containers/CodesplainAppBar.jsx]: I definitely need it for functional programming.

File: src/containers/CodesplainAppBar.jsx (hunk 1 above)
Comment (dane-johnson) [inline src/containers/CodesplainAppBar.jsx]: I'll add a comment, it builds a vector out of the nested objects that contains only the details relevant to the search menu, and is sortable by name.

File: src/containers/CodesplainAppBar.jsx (hunk 2 above)
Comment (solkaz) [inline src/containers/CodesplainAppBar.jsx]: I'm not saying you cannot use lodash, but rather that importing the default lodash object isn't allowed. Import only the functions you need.

File: src/containers/CodesplainAppBar.jsx [hunk 3]
```diff
@@ -7,6 +7,7 @@ import {
 } from 'material-ui';
 import { withRouter } from 'react-router';
 import cookie from 'react-cookie';
+import { reduce, concat, sortBy } from 'lodash';
```
Comment (solkaz) [inline src/containers/CodesplainAppBar.jsx]: Check the other files, like [here](https://github.com/maryvilledev/codesplainUI/blob/master/src/util/annotations.js#L1) as an example of how you should import lodash. Even though you're only importing a few functions from the default lodash object, you still have to import the entire thing, so it effectively is the same as before. Instead, it should be:
```js
import reduce from 'lodash/reduce';
/// ...
```

**Pull request merged**